zkevm-circuits/src/keccak_circuit/keccak_packed.rs

```diff
@@ -0,0 +1,1770 @@
+        .expect("Cannot parse DEGREE env var as usize")
+}
+
+fn get_mode() -> bool {
+    let mode: usize = var("MODE")
+        .unwrap_or_else(|_| "1".to_string())
+        .parse()
+        .expect("Cannot parse MODE env var as usize");
+    mode == 1
+}
+
+fn get_num_bits_per_lookup(range: usize) -> usize {
+    let num_unusable_rows = 31;
+    let degree = get_degree() as u32;
+    let mut num_bits = 1;
+    while range.pow(num_bits + 1) + num_unusable_rows <= 2usize.pow(degree) {
+        num_bits += 1;
+    }
+    num_bits as usize
+}
+
+fn get_num_bits_per_absorb_lookup() -> usize {
+    get_num_bits_per_lookup(ABSORB_LOOKUP_RANGE)
+}
+
+fn get_num_bits_per_theta_c_lookup() -> usize {
+    get_num_bits_per_lookup(THETA_C_LOOKUP_RANGE)
+}
+
+fn get_num_bits_per_theta_t_lookup() -> usize {
+    get_num_bits_per_lookup(THETA_T_LOOKUP_RANGE)
+}
+
+fn get_num_bits_per_rho_pi_lookup() -> usize {
+    if get_mode() {
+        get_num_bits_per_lookup(CHI_BASE_LOOKUP_RANGE.max(RHO_PI_LOOKUP_RANGE))
+    } else {
+        get_num_bits_per_lookup(RHO_PI_LOOKUP_RANGE)
+    }
+}
+
+fn get_num_bits_per_base_chi_lookup() -> usize {
+    if get_mode() {
+        get_num_bits_per_lookup(CHI_BASE_LOOKUP_RANGE.max(RHO_PI_LOOKUP_RANGE))
+    } else {
+        get_num_bits_per_lookup(CHI_BASE_LOOKUP_RANGE)
+    }
+}
+
+fn get_num_bits_per_ext_chi_lookup() -> usize {
+    get_num_bits_per_lookup(CHI_EXT_LOOKUP_RANGE)
+}
+
+/// AbsorbData
+#[derive(Clone, Default, Debug, PartialEq)]
+pub(crate) struct AbsorbData {
+    from: Word,
+    absorb: Word,
+    result: Word,
+}
+
+/// SqueezeData
+#[derive(Clone, Default, Debug, PartialEq)]
+pub(crate) struct SqueezeData {
+    packed: Word,
+}
+
+/// KeccakRow
+#[derive(Clone, Debug, PartialEq)]
+pub(crate) struct KeccakRow<F: Field> {
+    state: [F; KECCAK_WIDTH],
+    absorb_data: AbsorbData,
+    squeeze_data: SqueezeData,
+    cell_values: Vec<F>,
+    is_final: u64,
+    hash_rlc: F,
+}
+
+/// Part
+#[derive(Clone, Debug)]
+pub(crate) struct Part<F: Field> {
+    parts: Vec<Expression<F>>,
+    expr: Expression<F>,
+    num_bits: usize,
+}
+
+/// Part Value
+#[derive(Clone, Copy, Debug)]
+pub(crate) struct PartValue {
+    value: Word,
+    num_bits: usize,
+}
+
+/// KeccakConfig
+#[derive(Clone, Debug)]
+pub struct KeccakPackedConfig<F> {
+    q_enable: Selector,
+    q_first: Selector,
+    q_round: Column<Fixed>,
+    q_absorb: Column<Fixed>,
+    q_round_end: Selector,
+    is_final: Column<Advice>,
+    hash_rlc: Column<Advice>,
+    state: [Column<Advice>; KECCAK_WIDTH],
+    cell_values: Vec<Column<Advice>>,
+    absorb_from: Column<Advice>,
+    absorb_data: Column<Advice>,
+    absorb_result: Column<Advice>,
+    squeeze_packed: Column<Advice>,
+    round_cst: Column<Fixed>,
+    normalize_3: [TableColumn; 2],
+    normalize_4: [TableColumn; 2],
+    normalize_6: [TableColumn; 2],
+    chi_base_table: [TableColumn; 2],
+    chi_ext_table: [TableColumn; 2],
+    pack_unpack_table: [TableColumn; 2],
```

Comment:
Same as above with the `normalize` tables, except for the `pack_unpack_table` (a bit less optimized for now so just works on bytes, but could be improved), each of these tables are made as big as possible to pack in as much as possible. A tag I think will not improve things because then the tables can pack less bits together, and each cell in the circuit only ever needs to do a single kind of lookup so no need to be able to switch between them.